pkg/scheduler/framework/plugins/dynamicresources/dynamicresources.go

```diff
@@ -638,16 +910,138 @@ func (pl *dynamicResources) PreFilter(ctx context.Context, state *framework.Cycl
+				// This will get resolved by the resource driver.
+				return nil, statusUnschedulable(logger, "unallocated immediate resourceclaim", "pod", klog.KObj(pod), "resourceclaim", klog.KObj(claim))
+			}
+		}
+	}
+
+	if needResourceInformation {
+		// Doing this over and over again for each pod could be avoided
+		// by parsing once when creating the plugin and then updating
+		// that state in informer callbacks. But that would cause
+		// problems for using the plugin in the Cluster Autoscaler. If
+		// this step here turns out to be expensive, we may have to
+		// maintain and update state more persistently.
+		resources, err := newResourceModel(logger, pl.nodeResourceSliceLister, pl.claimAssumeCache)
+		if err != nil {
+			return nil, statusError(logger, err)
 		}
+		s.resources = resources
 	}
 
 	s.claims = claims
-	state.Write(stateKey, s)
 	return nil, nil
 }
 
+func (pl *dynamicResources) lookupParameters(logger klog.Logger, class *resourcev1alpha2.ResourceClass, claim *resourcev1alpha2.ResourceClaim) (classParameters *resourcev1alpha2.ResourceClassParameters, claimParameters *resourcev1alpha2.ResourceClaimParameters, status *framework.Status) {
+	classParameters, status = pl.lookupClassParameters(logger, class)
+	if status != nil {
+		return
+	}
+	claimParameters, status = pl.lookupClaimParameters(logger, claim)
+	return
+}
+
+func (pl *dynamicResources) lookupClassParameters(logger klog.Logger, class *resourcev1alpha2.ResourceClass) (*resourcev1alpha2.ResourceClassParameters, *framework.Status) {
+	if class.ParametersRef == nil {
+		return nil, nil
+	}
+
+	if class.ParametersRef.APIGroup == resourcev1alpha2.SchemeGroupVersion.Group &&
+		class.ParametersRef.Kind == "ResourceClassParameters" {
+		// Use the parameters which were referenced directly.
+		parameters, err := pl.classParametersLister.ResourceClassParameters(class.ParametersRef.Namespace).Get(class.ParametersRef.Name)
+		if err != nil {
+			if apierrors.IsNotFound(err) {
+				return nil, statusUnschedulable(logger, fmt.Sprintf("class parameters %s not found", klog.KRef(class.ParametersRef.Namespace, class.ParametersRef.Name)))
+			}
+			return nil, statusError(logger, fmt.Errorf("get class parameters %s: %v", klog.KRef(class.Namespace, class.ParametersRef.Name), err))
+		}
+		return parameters, nil
+	}
+
+	allParameters, err := pl.classParametersLister.ResourceClassParameters(class.Namespace).List(labels.Everything())
+	if err != nil {
+		return nil, statusError(logger, fmt.Errorf("listing class parameters failed: %v", err))
+	}
+	for _, parameters := range allParameters {
+		if parameters.GeneratedFrom == nil {
+			continue
+		}
+		if parameters.GeneratedFrom.APIGroup == class.ParametersRef.APIGroup &&
+			parameters.GeneratedFrom.Kind == class.ParametersRef.Kind &&
+			parameters.GeneratedFrom.Name == class.ParametersRef.Name &&
+			parameters.GeneratedFrom.Namespace == class.ParametersRef.Namespace {
+			return parameters, nil
+		}
+	}
```

Comment:
The E2E tests cover all three scenarios (classic DRA, directly referencing in-tree types, generated in-tree types) for the same test cases. I think this is low risk and important to have.

The CEL expression might be non-trivial to write manually and lacks vendor-specific validation. We've discussed this quite a bit during the KEP phase and agreed that supporting vendor CRD parameters for class and claim is useful or even required (depending whom you ask).

Reply:
The expectation is that ResourceClasses most often will be installed along with the driver (i.e., written by the vendor). Admins could also write them. As Patrick said, there is something of a debate on the details here and this is a safe starting point.

Reply:
Ok, in that case, we are going to need some form of Index, instead of listing every object (are they always namespaced?). However, we can leave that for the next release.

Reply:
Okay.

Reply:
cc @thockin, since you mentioned a dislike for this API